bin/ch-completion.bash

```diff
@@ -0,0 +1,388 @@
+#   foo bar baz
+#   $ echo ${foo[@]:1:3}
+#   bar baz qux
+#
+# FIXME: Add syntax glossary
+#
+
+# Possible extensions once this is merged:
+#   * Add completion support for non-bash shells (e.g. zsh and tchs).
+#     see https://github.com/git/git/tree/master/contrib/completion
+#   * Add support for mid-line completion (possibly using COMP_POINT,
+#     see https://devmanual.gentoo.org/tasks-reference/completion/index.html).
+#
+
+# FIXME: disable shellcheck SC2034 for this? (See base.sh)
+CH_BIN="$(cd "$(dirname -- "${BASH_SOURCE[0]}")" && pwd)"
+
+# Debugging log
+if [[ -f "/tmp/ch-completion.log" ]]; then
+    rm /tmp/ch-completion.log
+fi
+
+## ch-image ##
+
+# Subcommands and options for ch-image
+#
+_image_build_opts="-b --bind --build-arg -f --file --force
+                   --force-cmd -n --dry-run --parse-only -t --tag"
+
+_image_common_opts="-a --arch --always-download --auth --cache
+                    --cache-large --dependencies -h --help
+                    --no-cache --no-lock --profile --rebuild
+                    --password-many -s --storage --tls-no-verify
+                    -v --verbose --version"
+
+_image_subcommands="build build-cache delete gestalt
+                    import list pull push reset undelete"
+
+## ch-image ##
+
+# Completion function for ch-image
+#
+_ch_image_completion () {
+    local prev
+    local cur
+    local words
+    local sub_cmd
+    local strg_dir
+    local extras=""
+    _get_comp_words_by_ref -n : cur prev words
+
+    # If "$cur" is non-empty, we want to ignore it as a potential subcommand
+    # to avoid unwanted behavior. “${words[@]::${#words[@]}-1}” gives you all
+    # but the last element of the array “words” (see syntax glossary above) so 
+    # if "$cur" is non-empty, pass it to _ch_subcommand_get. Otherwise pass all
+    # elements of “words.”
+    if [[ -n "$cur" ]]; then
+        sub_cmd=$(_ch_subcommand_get "$_image_subcommands" "${words[@]::${#words[@]}-1}")
+        strg_dir=$(_ch_find_storage "${words[@]::${#words[@]}-1}")
+    else
+        sub_cmd=$(_ch_subcommand_get "$_image_subcommands" "${words[@]}")
+        strg_dir=$(_ch_find_storage "${words[@]}")
+    fi
+
+    # Common opts that take args
+    #
+    case "$prev" in
+    -a | --arch )
+        # FIXME: Remove yolo?
+        # FIXME: Missing common architectures?
+        COMPREPLY=( $(compgen -W "host yolo 386 amd64 arm/v6 
+                                  arm/v7 arm64/v8 ppc64le s390x" -- $cur) )
+        return 0
+        ;;
+    --cache-large )
+        # This is just a user-specified number. Can't autocomplete
+        COMPREPLY=()
+        return 0
+        ;;
+    -s | --storage )
+        # This “if” helps avoid overzealous completion. E.g. if there's
+        # only one subdir of the current dir, this command completes to
+        # that dir even if "$cur" is empty. I didn't like that, hence the
+        # “if”.
+        if [[ -n "$cur" ]]; then
+            compopt -o nospace
```

Comment:
Does setting this option persist to future calls of `_ch_image_completion`? If so, should probably set default options at top of the function.

Reply:
My testing suggests it doesn't persist